### PR TITLE
Core/Movement: Rewrite confused movement

### PR DESCRIPTION
**Pull request by Mihc3**
Based on `15595_2012-08-22_22-02-32_looking_for_dungeon_delete_item_set.pkt`.

I gathered waypoints and made a nice graph/map of a few confusion effects:
https://www.desmos.com/calculator/cipr26agwb

**Changes proposed:**
- Unit will now make 5 moves and not 2 (during confusion that lasts 4 seconds).
    - New waypoint calculated right after 800ms, no wait time.
- Distance from original point seems to be more or less rounded value.
- Less unnecessary IF statements and indentation.

**Target branch(es):** 3.3.5/master

- 3.3.5 (probably?)
- master ✔️

**Issues addressed:** Closes #  (insert issue tracker number)

**Tests performed:** (Does it build, tested in-game, etc.)
* Build: ✔️
* Tests:
    * Confusion (41397) cast on enemy creatures nearby ✔️
    * Confusion (170349) cast on self ✔️

**Known issues and TODO list:** (add/remove lines as needed)
* None.

### CODE REVIEW

File: src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp [hunk 1]
```diff
@@ -56,44 +56,38 @@ bool ConfusedMovementGenerator<T>::DoUpdate(T* unit, uint32 diff)
     if (unit->HasUnitState(UNIT_STATE_ROOT | UNIT_STATE_STUNNED | UNIT_STATE_DISTRACTED))
         return true;
 
-    if (i_nextMoveTime.Passed())
-    {
-        // currently moving, update location
-        unit->AddUnitState(UNIT_STATE_CONFUSED_MOVE);
```
Comment (joschiwald) [inline src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp]: what happend with this unit state?

Comment (Mihapro) [reply src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp]: Isn't adding it on init and reset enough? Or does something remove it?

Comment (Mihapro) [reply src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp]: It doesn't seem anything in core would clear all unit states or specifically confused move - only confused movement generator itself removes it

File: src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp [hunk 2]
```diff
@@ -31,73 +31,76 @@ void ConfusedMovementGenerator<T>::DoInitialize(T* unit)
...
+
+    // Check how far we can move (collision check).
+    unit->MovePositionToFirstCollision(pos, dist, angle);
+
+    // Generate path
+    PathGenerator path(unit);
+    path.SetPathLengthLimit(10.0f);
+    bool result = path.CalculatePath(pos.m_positionX, pos.m_positionY, pos.m_positionZ);
+    if (!result || (path.GetPathType() & PATHFIND_NOPATH))
+        return true; // Return without setting timer, generate another path on next update.
```
Comment (tkrokli) [inline src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp]:
```cpp
src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp:88:9:
 fatal error: void function '_move' should not return a value [-Wreturn-type]

        return true; // Return without setting timer, generate another path on next update.
        ^      ~~~~
```
(`_move` in Line 70 - https://github.com/TrinityCore/TrinityCore/pull/20788/files#diff-6924fe8c6e57209e1beffb3de0491242R70)

**Comment by jackpoz** [comment]
this PR has a lot of changes about "Less unnecessary IF statements and indentation." that makes it harder to review. could you just include the bare minimum changes required ?

**Comment by Mihapro** [comment]
I only changed from ```if/else``` to ```if(not)->return``` so it doesn't need to check timer twice in one update (timer.passed() thing), I didn't touch most of the code 😄

**Comment by ccrs** [comment]
this doesnt apply to 335, its already implemented and most likely improved

**Comment by jackpoz** [comment]
So master just needs some cherrypicks ?

**Comment by MaxtorCoder** [comment]
> 
> 
> So master just needs some cherrypicks ?

Yes, definitely :P

**Comment by jackpoz** [comment]
Closing this PR then

**Pull request closed without merging**